### PR TITLE
switch to auto-increment ResourceId

### PR DESCRIPTION
**Pull request by bmacer**
addresses #138 
One issue that remains is that id is part of the `Resource` struct.  I don't think this is necessary generally.  For example, Nft and Collection don't have their respective IDs as part of their struct.  However, we need to extract the resource ID when doing an iter_prefix_values for the CollectionID, NftId inside the `do_equip` function (we do this when checking for the existence of a BaseId+SlotId on the resources for a Slot Resource -- that is, ensuring that there exists a Slot Resource matching the BaseId and SlotId passed to the equip extrinsic).

https://github.com/rmrk-team/rmrk-substrate/blob/218fe30a8950ab0acf51976246fcd6b812aa8a87/pallets/rmrk-equip/src/functions.rs#L255-L268

We could either figure out how to get the key value when iterating through with `iter_prefix_values` (I asked the question in StackExchange [here](https://substrate.stackexchange.com/questions/2767/get-value-and-key-from-iter-prefix-values)), or find a more graceful way of checking altogether.

For now, holding the `id` in the struct works.

### CODE REVIEW

**Review by HashWarlock — APPROVED** [review]
Overall looks good.

**Comment by HashWarlock** [comment]
> addresses #138 One issue that remains is that id is part of the `Resource` struct. I don't think this is necessary generally. For example, Nft and Collection don't have their respective IDs as part of their struct. However, we need to extract the resource ID when doing an iter_prefix_values for the CollectionID, NftId inside the `do_equip` function (we do this when checking for the existence of a BaseId+SlotId on the resources for a Slot Resource -- that is, ensuring that there exists a Slot Resource matching the BaseId and SlotId passed to the equip extrinsic).
> 
> https://github.com/rmrk-team/rmrk-substrate/blob/218fe30a8950ab0acf51976246fcd6b812aa8a87/pallets/rmrk-equip/src/functions.rs#L255-L268
> 
> We could either figure out how to get the key value when iterating through with `iter_prefix_values` (I asked the question in StackExchange [here](https://substrate.stackexchange.com/questions/2767/get-value-and-key-from-iter-prefix-values)), or find a more graceful way of checking altogether.
> 
> For now, holding the `id` in the struct works.

I think the `do_equip` function handles too much and can be confusing for downstream in the case. We can end up making this function easier to handle by decoupling the unequip logic into its own flow. I'll try to think of a way to better optimize the storage logic of resources. I have a couple questions:

- Do Resources have the ability to be more than just a Base(Composable), Slot or Media(Basic)?
  - If that's the case then 2 resource types would be able to be assigned a priority Base & Media, right?

I think if we separate these into their own structs then have an enum for the ResourceType & then make the structs optionals in the ResourceInfo struct. I'll create an issue & try to create a diagram & code to explain what it would look like.

**Review by ilionic — COMMENTED** [review]
Since iteration over current Resources storage is suboptimal ( as pointed out by Shawn in the linked comment ) we should consider switching it to BoundedVec part of NFT / Base itself so it's safe.

File: traits/src/resource.rs [hunk 1]
```diff
@@ -108,57 +108,49 @@ pub enum ResourceTypes<BoundedString, BoundedParts> {
 	Slot(SlotResource<BoundedString>),
 }
 
-
 #[derive(Encode, Decode, Eq, PartialEq, Clone, RuntimeDebug, TypeInfo, MaxEncodedLen)]
 #[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
-pub struct ResourceInfo<BoundedResource, BoundedString, BoundedParts> {
```
Comment (HashWarlock) [inline traits/src/resource.rs:114]: Couldn't we keep `src`, `metadata`, `license` & `thumb` in here? I was thinking `ResourceInfo` would have the `ResourceType` that we could `match` on then get the contents from the nested struct then handle logic from there.

File: traits/src/resource.rs [hunk 2]
```diff
@@ -108,57 +108,49 @@ pub enum ResourceTypes<BoundedString, BoundedParts> {
...
 
 /// Abstraction over a Resource system.
-pub trait Resource<BoundedString, AccountId, BoundedResource, BoundedPart> {
+pub trait Resource<BoundedString, AccountId, BoundedPart> {
 	fn resource_add(
 		sender: AccountId,
 		collection_id: CollectionId,
 		nft_id: NftId,
-		resource_id: BoundedResource,
 		resource: ResourceTypes<BoundedString, BoundedPart>,
```
Comment (ilionic) [inline traits/src/resource.rs:136]: minor thing, purely readability - how renaming `resource` to more explicit ( ie resource_type )

**Review by ilionic — COMMENTED** [review]
Looking good, much cleaner. I think that's an acceptable tradeoff having additional storage to avoid unbounded iterations but curious to hear more opinions.

**Comment by ilionic** [comment]
Merging to avoid conflicts. @HashWarlock feel free to extract extra comments to separate issue. @bmacer CC

**Pull request merged**